test/builtin_dataset_mocks.py

```diff
@@ -72,6 +74,25 @@ def prepare(self, home, config):
         available_file_names = {path.name for path in root.glob("*")}
         required_file_names = {resource.file_name for resource in self.dataset.resources(config)}
         missing_file_names = required_file_names - available_file_names
+        extra_file_names = available_file_names - required_file_names
+
+        # Some datasets need to provide already preprocessed data, for example if the preprocessing includes downloads
+        if extra_file_names:
+            for extra in extra_file_names.copy():
+                candidate_pattern = re.compile(fr"^{extra.split('.', 1)[0]}([.]\w+)*$")
+                try:
+                    missing = next(missing for missing in missing_file_names if candidate_pattern.match(missing))
+                except StopIteration:
+                    continue
+
+                extra_file_names.remove(extra)
+                missing_file_names.remove(missing)
```

Comment:
I've expanded the comment above the block. LMK if it is sufficient.

Reply:
Sorry I still don't understand the regex (do we really need a regex?), not why the raw data ends up in missing_file_names.

The problem we're having seems reasonably simple, yet the code to handle it looks fairly complex. Is there something we can do to simplify the logic here?

Reply:
> why the raw data ends up in missing_file_names.

The `required_file_names` include the name of the resources defined on the dataset, while the `available_file_names` just glob the directory. So if the dataset defines `OnlineResource(file_name="foo.tar")`, but we only find a folder `foo`, `foo.tar` will be in `missing_file_names` while `foo` will be in `extra_file_names`. This code block finds this correspondence and subsequently removes both from the sets.

> Is there something we can do to simplify the logic here?

Will try.

Reply:
In the example you describe, `foo.tar` was removed during the "preprocessing" stage? Or just not created at all in the mock because we don't actually need it?

Reply:
Let's make it more concrete: the SBU dataset provides a `SBUCaptionedPhotoDataset.tar.gz` archive, which contains a list of URLs for images and corresponding captions. Thus, before you can use the dataset, we need download all of these images. In #5683 we opted to put this functionality into the `preprocess` stage. Thus, when the dataset is loaded the first time, the archive is downloaded and the preprocessing extracts it and starts the download of all the images. Since we are not repackaging the images, we put the original data as well as the downloaded images into a `SBUCaptionedPhotoDataset` folder. With this, our automatic resource loading automatically picks the extracted folder to load the resource and thus also load the images.

> In the example you describe, `foo.tar` was removed during the "preprocessing" stage? Or just not created at all in the mock because we don't actually need it?

TL;DR: `foo.tar` is not created by the test suite.

In the test suite, we cannot perform the preprocessing, because it would require internet access and would take quite some time even if we provide fake images (this is what @lezwon had as a first draft). Thus, the easiest way to provide mock data is to provide the folder `SBUCaptionedPhotoDataset` directly rather than `SBUCaptionedPhotoDataset.tar.gz`. Thus, if the dataset loads the resource it is tricked into loading the "already preprocessed folder" rather than the raw archive.

Reply:
Would things be simpler if the extra download wasn't part of the `preprocess` stage?

Reply:
It would make the test suite easier yes, but the resource loading more complicated. Where would you put this logic? So far this is the only dataset that has a setup like this.

Reply:
> Where would you put this logic?

Could we do it after the resource is loaded, withing the `make_dp()` method? 

> So far this is the only dataset that has a setup like this.

Indeed, this is why it's worth weighing whether it's worth special-casing the test suite just for this single dataset.

Reply:
Not sure what you mean here? Do you want a general solution on dataset side? Because for the test suite this PR is a general solution for all datasets that have a similar setup. Nothing is special about SBU here.

Reply:
> this PR is a general solution for all datasets that have a similar setup

How many datasets have a similar setup? From above it seems like this is the only one. Hence I'm wondering whether having this custom logic, on which you and I have spent a few hours, already is really necessary.

Reply:
Prototype for loading inside `_make_dp` is in lezwon/vision#1.

Reply:
> How many datasets have a similar setup? From above it seems like this is the only one.

Nope, at least Kinetics also has a similar setup:

https://github.com/pytorch/vision/blob/095cabb76cdcd4763bad629481d189b91e3df42c/torchvision/datasets/kinetics.py#L181-L190

> Hence I'm wondering whether having this custom logic, on which you and I have spent a few hours, already is really necessary.

What is the alternative? Dropping support for the dataset? We might be able to argue that SBU is not important enough to "break our rules", but Kinetics certainly is.